spec.bs

```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. Let |outsideSettingsOrigin| be |environment|'s [=environment settings object/origin=].
+    1. If |outsideSettingsOrigin| is an [=opaque origin=], then return false.
     1. If |origin| is an [=opaque origin=], then return false.
     1. Let |globalObject| be the [=current realm=]'s [=global object=].
     1. [=Assert=] that |globalObject| is a {{Window}} or a {{SharedStorageWorkletGlobalScope}}.
-    1. If |globalObject| is a {{Window}} and |globalObject|'s [=associated document=] is not [=allowed to use=] the "[=PermissionsPolicy/shared-storage=]" feature, return false.
+    1. If |globalObject| is a {{Window}}, and if the result of running [=Is feature enabled in document for origin?=] on "[=PermissionsPolicy/shared-storage=]", |globalObject|'s [=associated document=], and |origin| returns false, then return false.
+    1. Return true.
+  </div>
+
+  <div algorithm>
+    To <dfn>determine whether shared storage is allowed by enrollment and user preference</dfn>, given an [=environment settings object=] |environment| and an [=url/origin=] |origin|, run these steps:
+
     1. If the result of running [=obtaining a site|obtain a site=] with |origin| is not [=enrolled=], then return false.
-    1. If the result of running [=check if user preference setting allows access to shared storage=] from |environment| is false, then return false.
+    1. If the result of running [=check if user preference setting allows access to shared storage=] given |environment| and |origin| is false, then return false.
     1. Return true.
   </div>
 
+  <div class="note">
+    Here are the scenarios where the algorithms [=determine whether shared storage is allowed by context=] and [=determine whether shared storage is allowed by enrollment and user preference=] could be used:
+
+    - For <a href="#window-setter">Window setter/deleter methods</a>, |environment|'s [=environment settings object/origin=] should equal |origin|.
+    - For creating a worklet, and for initiating (from {{Window}}) and running (from {{SharedStorageWorkletGlobalScope}}) operations on a worklet, |environment| should be the worklet's creator window, and |origin| should be the [=SharedStorageWorklet/worklet origin=].
+    - For a <a href="#ss-fetch-algo">fetch request</a>, |environment| should be the request's initiator window, and |origin| should be the request's URL's origin.
+  </div>
+
   <div algorithm>
-    To <dfn>check if addModule is allowed and update status</dfn> for a {{SharedStorageWorklet}} |worklet|, run the following steps:
-    1. If the result of running [=determine whether shared storage is allowed=] on the [=relevant settings object=] of [=this=] is false, return false.
-    1. If |worklet|'s [=addModule initiated=] is true, return false.
+    To <dfn>check if addModule is allowed and update status</dfn> given a {{SharedStorageWorklet}} |worklet| and an [=/URL=] |moduleURL|, run the following steps:
+    1. Let |workletOrigin| be |moduleURL|'s [=url/origin=].
```

Comment:
This is interesting, but perhaps a little bit worrying. I want to check something, but let me know if I am missing something. At some point in the worklet creation process we need to call https://html.spec.whatwg.org/C#set-up-a-worklet-environment-settings-object, which actually sets up the global and its settings object appropriately. This is responsible for setting the global's origin, which is presumably what you _actually_ want to set `|workletOrigin|` to, right? Is this ever being done outside of this algorithm, or are we just missing it?

Reply:
It seems the worklet will set it to the expected value during [addModule](https://html.spec.whatwg.org/multipage/worklets.html#dom-worklet-addmodule)() during its "Create a worklet global scope" step, so I assume there's no need to additionally mention it in this spec?

This `workletOrigin` member may seem redundant. But at this point, we do check for user preferences at the early stage of addModule(), so it also seems okay to have this member explicitly for that check (and for others checks as a convenience).

Reply:
OK so if I understand correctly, the intention is for https://pr-preview.s3.amazonaws.com/WICG/shared-storage/pull/131.html#sharedstorageworklet-worklet-origin to match, identically, the ESO's origin? If that's the case, can you add a note about this underneath the definition?

----

Also I want to point out one thing:

> 1. Let |workletOrigin| be |moduleURL|'s [=url/origin=].

Note that this is happening too early. The URL has been parsed yet, so `moduleURL` is just a string and doesn't have the concept of an "origin". This has to run _after_ this line: https://html.spec.whatwg.org/multipage/worklets.html#dom-worklet-addmodule:~:text=If%20moduleURLRecord%20is%20failure%2C%20then%20return%20a%20promise%20rejected%20with%20a%20%22SyntaxError%22%20DOMException